### PR TITLE
Add native support for Criteo adapter

### PR DESCRIPTION
**Pull request by Swiiip**
## Type of change
- [ x] Feature

## Description of change
This feature provides a new parameter for the Criteo adapter:
New parameter: nativeCallback
Type: javascript function
Parameter: criteo native json response

This callback is called on the returned json for native zones.

### CODE REVIEW

**Comment by Swiiip** [comment]
Hello guys, could we have an update on this PR please ?
Thanks

**Comment by mkendall07** [comment]
@Swiiip 
Do you work for Criteo? 
@icaroulle any thoughts on this?

**Comment by Swiiip** [comment]
@mkendall07 Yes I do work for Criteo

**Comment by icaroulle** [comment]
@Swiiip aka Hugo h.duthil@criteo.com is indeed working for Criteo. :smile: 
@mkendall07 Can you please review this change?

**Comment by mkendall07** [comment]
Ok thanks for the info guys. @matthewlane can you review?

**Comment by mkendall07** [comment]
@Swiiip 
Can you post an example function that would be passed to `nativeCallback` ?

**Comment by Swiiip** [comment]
@mkendall07 

An example of native callback: 

```
function(json) { 
    var li = $(".ccr-latest-post li")[0];
    var htmlcode = '';
    htmlcode += '<div class="ccr-thumbnail">';
    htmlcode += '<img src="'+ json.privacy.optout_image_url + '" style="width:19px;height:15px;position:absolute;right: 0px">';
    htmlcode += '<img src="'+ json.advertiser.logo.url + '" style="position: absolute; top: 0; left: 7px; padding: 3px 2px; line-height: 1; margin-left: -15px; font-size: 7px; letter-spacing: 1px; display: inline-block; text-transform: uppercase; -webkit-transform: rotate(-15deg); -moz-transform: rotate(-15deg); -ms-transform: rotate(-15deg); -o-transform: rotate(-15deg); transform: rotate(-15deg); width: 50% ">';
    htmlcode += '<img src="' + json.products[0].image.url + '" alt="Thumbnail 1" width="239px" height="191px" >';
    htmlcode += '<p><a href="' +  json.products[0].click_url + '">'+ "Go to: " + json.advertiser.domain +'</a></p>';
    htmlcode += '</div>';
    htmlcode += '<h4><a href="' + json.products[0].click_url + '">' + json.products[0].description.substring(0,40) + '</a></h4>';
    htmlcode += '<img src="' +json.products[0].viewurl + '" width="0px" height="0px" style="display:none" >';
    for (i in json.impression_pixels) {
        htmlcode += '<img src="' + json.impression_pixels[i].url + '" width="0px" height="0px" style="display:none" >';
    }							
    li.innerHTML = htmlcode;
    console.log(json);
}
```

**Comment by allanjun** [comment]
hi @mkendall07 any updates on this ?

File: src/adapters/criteo.js [hunk 1]
```diff
@@ -134,6 +123,50 @@ var CriteoAdapter = function CriteoAdapter() {
...
+        } else {
+
+          // store the callbacks in a global object
+          window.criteo_pubtag.native_slots = window.criteo_pubtag.native_slots || {};
+          window.criteo_pubtag.native_slots[""+bidObject.adId] = { callback: slot.nativeCallback, nativeResponse: bidResponse.native };
+
+          // this code is executed in an iframe, we need to get a reference to the
+          // publishertag in the main window to retrieve native responses and callbacks.
+          // it doesn't work with safeframes
+          bidObject.ad =  "<script type=\"text/javascript\">";
```
Comment (mkendall07) [inline src/adapters/criteo.js]: please use string literals for constructing the templates. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

**Comment by mkendall07** [comment]
@allanjun @Swiiip please rebase to resolve conflicts and update to use string templates then we can merge. Also would be great to get the API documented. Thanks

File: src/adapters/criteo.js (hunk 1 above)
Comment (Swiiip) [inline src/adapters/criteo.js]: done 👍

**Comment by Swiiip** [comment]
I updated the PR, we will provide updated documentation for the adapter once validated internally.

**Comment by muuki88** [comment]
> we will provide updated documentation for the adapter once validated internally.

That's greate news @Swiiip  :heart_eyes: We currently read the code to figure out how to configure prebid, which is kinda awkward ;)

File: test/spec/modules/criteoBidAdapter_spec.js [hunk 2]
```diff
@@ -157,11 +175,42 @@ describe('criteo adapter test', () => {
...
+      server = sinon.fakeServer.create({ autoRespond: true, respondImmediately: true });
+      server.respondWith(JSON.stringify(validNativeResponse));
+    });
+
+    it('adds creative to the response of a native valid request', (done) => {
+      stubAddBidResponse = sinon.stub(
+        bidManager, 'addBidResponse',
+        function (adUnitCode, bid) {
+          let expectedAdProperty = `<script type=\"text/javascript\">
+  let win = window;
```
Comment (mkendall07) [inline test/spec/modules/criteoBidAdapter_spec.js]: might consider indenting these lines.

**Comment by mkendall07** [comment]
@Swiiip it looks like the build is failing due to linting errors. Can you rebase 1 more time please and run `eslint --fix` on your file (or use your favorite editor to lint fix).

**Comment by Swiiip** [comment]
All set :)

**Pull request merged**